### ISSUE TITLE
Introduce extension API for providing a custom ClassLoader (e.g., for Powermock)

### ISSUE THREAD

**Issue by mepeisen** [junit-team/junit5#201]
## Overview

I tried to build a Powermock `Extension` similar to the Mockito example. I have to test some classes that create new objects within their constructors. In JUnit 4 I simply use the `PowerkmockRunner` and the `whenNew` from Powermock.

At least I ended with many problems and gave up. However in JUnit 4 Powermock creates an instrumented class loader and even duplicates everything (the test instance etc.) by using the new class loader. It is simple because the loader/rule are aware of executing the test itself. In JUnit Jupiter an `Extension` is not able to execute the test by itself.

I am wondering how this will be done in JUnit Jupiter. Are you planning to introduce some `CreateTestInstance` extension API? Something that is aware of manipulating class loaders. Or at least an `AroundAll` extension API to instrument the whole test class?

## Related Issues

- #157
- #805
- #806 
- #1680
- #1799

**Comment by jlink** [comment]
What we have is the `InstancePostProcessor` extension point. Currently, it does not allow to return a different test instance but we have been discussing to allow that, e.g.:

```
public interface InstancePostProcessor extends ExtensionPoint {
        /** Return the original or a wrapped instance of the test object */
	Object postProcessTestInstance(TestExtensionContext context) throws Exception;
}

```
Would that be sufficien for your purposes? I am curious though, why PowerMock has to change the test instance (which can easily lead to incompatibilities with other extensions). Maybe there's another way to go about it?

As for an extension point that wraps test execution itself, we are currently shying away from it for reasons discussed (among other places) in #157.

**Comment by mepeisen** [comment]
As far as I can see returning a different instance should fit the needs.
I do not know why powermock needs it but I will ask the authors. All I can see is that mocking objects and classes already works.

Actually in my Scenario i need to mock things happening in constructors.

`class ToTest {
public ToTest() {
    this.aObj = new A();
    doSomeThingWithAObj();
}
}`

And I need to inject a mock for "aObj". Actually aObj is a console reader to receive commands.

Mabye exactly this scenario need playing around with class loaders and replacing the test instance. I will invite the powermock guys to discuss this in this issue :-)

**Comment by jlink** [comment]
2016-03-14 8:43 GMT+01:00 Martin Eisengardt <notifications@github.com>:

> Mabye exactly this scenario need playing around with class loaders and
> replacing the test instance. I will invite the powermock guys to discuss
> this in this issue :-)


Thanks. This should certainly help :)

**Comment by thekingnothing** [comment]
@mepeisen, thanks for raising the problem. We ware going to look to jUnit5 and implementing supporting of jUnit5 after we finish preparing current release. And maybe it could be to late :) 

PowerMock uses custom class loader to modified loaded class and either remove private/final modified or modify body of static classes to add call interceptors. As result instance which created via `mock` `@Mock` could have classes which is loaded by custom class loader and cannot be casted to classes which are loaded by test class loader. What's why PowerMock has to change the test instance and reload it via custom class loader. 

By the way, PowerMock creates a new instance of PowerMock class loader for test chunk. Test chunk is a set of tests which have to be run with same set of modified classes and declared via `@PrepareForTest` annotation. This annotation could be used either on class or on method level. 

Example: 

```
TestClass{

     @PrepareForTest(SomeStaticClass.class)
     public void testA()
     }

     @PrepareForTest(SomePrivateClass.class)
     public void testB()
     }

     public void testC()
     }
}
```

In this case PowerMock creates two test chunk and two class loader: one for running test `testA` and another to run `TestB`. The `TestC` is run with either first class loader or  second depends on order in array which is returned by `testClass.getMethods()`.

When a test is run with jUnit4 Runner then some jUnit logic is duplicated in our code and it's not very nice and I prefer to avoid it.

**Comment by thekingnothing** [comment]
Good example, how we can avoid over controlling - TestNG `ObjectFactory`.  This class has a method to create a new instance of class. PowerMock create a new instance, which is loaded via PowerMock class loader and wrapped by proxy, to clear PowerMock internals after each test. 

We also try to use Java Agent to get off necessity using custom class loader, but Java Instrumentation limits us by allowing only modify body of method.

**Comment by thekingnothing** [comment]
@johanhaleby what do think?

**Comment by mepeisen** [comment]
Clearing internal states can already be done via AfterEach-ExtensionPoint or am I wrong?

**Comment by thekingnothing** [comment]
@mepeisen, I don't know how it could be done in jUnit5. I spoke about jUnit4 (where when we use runner we have to control test execution or if use Rule then reload all classes with PowerMock class loader) and about TestNG (where not way to programatically register listeners, so we emulate listeners by using cglib). 

It'll be nice if jUnit5 give us ability to control test instance creating and programatically add `ExtensionPoint` to avoid asking a user add some additional annotation or make any additional action. 

For example. We ask a user register somehow a PowerMock implementation of `ExtensionPoint` and all over stuff will be done by PowerMock.

**Comment by mepeisen** [comment]
I will create an example by trying to rebuild the powermock runner. I already did it locally but ran into multiple problems trying to replace the test instance. Maybe I did something illegal with powermock :-)

**Comment by jlink** [comment]
As already said. There is currently no way to replace test instance in
JUnit5.

2016-03-14 14:30 GMT+01:00 Martin Eisengardt <notifications@github.com>:

**Comment by thekingnothing** [comment]
@jlink, I think, that suggested approach with be sufficient for PowerMock, in case if there will be a way to reloaded test Instance via PowerMock class loader, for example by making deep copy or make a new instance by using `Class.forName. I mean approach with:

```
public interface InstancePostProcessor extends ExtensionPoint {
        /** Return the original or a wrapped instance of the test object */
    Object postProcessTestInstance(TestExtensionContext context) throws Exception;
}
```

**Comment by jlink** [comment]
I created an issue to implement that (#203) and blocked this issue until #203 will be resolved.

**Comment by johanhaleby** [comment]
@thekingnothing describes what PowerMock is doing today and it gets complex because of the so called "test chunking" feature. Just like @thekingnothing says you have the ability to execute an individual test method with a new classloader. What actually happens under the cover is that PowerMock creates a new JUnit test for each test chunk (loaded by an individual classloader) and creates an "uber test" that calls all child tests. To the end user it will look like only one test is executed but it may actually be many tests making up the "uber test". 

But this is (as you can imagine) very complex and it's a feature that I don't think is used very much (and I think that actual use case for it is very slim). So I don't mind dropping this feature from PowerMock when it comes to JUnit 5. 

What we do need though is a way to hook in our classloader. Doing a deep copy is very slow (and error prone, at least when it comes to copy static final fields between classloaders) so if we can avoid doing that it would be nice. If I remember things correctly (I think @thekingnothing is more up to date on this) PowerMock currently loads JUnit from our classloader (i.e. PowerMock is driving the test execution) but from what I understand JUnit 5 will drive PowerMock? I wonder if this will work with classloaders? 

PowerMock also has an agent which might work for this but the problem with agents are that you need to start them explicit (well, without hacks anyways) which defines the purpose of PowerMock. Agents are also more limited than classloaders and you can't do things like suppress static initializers etc afaik.

**Comment by mepeisen** [comment]
I finally got a working Showcase :-)

Classic JUnit 4 variant:
https://github.com/mepeisen/xw-nukkit-test/blob/master/src/test/java/eu/xworlds/nukkit/spowermock/Junit4Test.java

JUnit5 variant without powermock standalone:
https://github.com/mepeisen/xw-nukkit-test/blob/master/src/test/java/eu/xworlds/nukkit/spowermock/JUnit5Test.java

And now a Show case where my custom Extension injects "prepareForTest":
https://github.com/mepeisen/xw-nukkit-test/blob/master/src/test/java/eu/xworlds/nukkit/servertests/StopCommandTest.java
The injection is done here: https://github.com/mepeisen/xw-nukkit-test/blob/master/src/main/java/eu/xworlds/nukkit/test/NukkitExtension.java#L58-L63

The showcase powermock Extension can be found here:
https://github.com/mepeisen/xw-nukkit-test/blob/master/src/main/java/eu/xworlds/nukkit/test/sample/PowermockExtension.java
Of Course not very Feature rich. Many annotations are not supported to make the showcase simple. But it works. :-)

However as long as #203 is not implemented the "returing new test instance" is a very bad hack: https://github.com/mepeisen/xw-nukkit-test/blob/master/src/main/java/eu/xworlds/nukkit/test/sample/PowermockExtension.java#L140-L144

**Comment by mepeisen** [comment]
This lead to a new question: I now have an Extension (NukkitExtension) that depends on another Extension (PowermockExtension). Currently the test code must ensure that both extensions are loaded and that they even are in correct order.
Are there any plans to describe such dependencies and sort orders?

**Comment by jlink** [comment]
@mepeisen 

To make sure that two extensions are loaded together, just introduce a meta annotation that loads both (see http://junit-team.github.io/junit5/#annotations section 3.1.1). I'm not sure that's a good approach for most cases. If two extensions must go together, just make it one - maybe through using a registrar (see http://junit-team.github.io/junit5/#extension-registration). Individual extensions should be independent from each other.

Ordering of extensions is possible (also using a registrar); It's badly documented though (on purpose), since this mechanism will probably change fundamentally till milestone 1 or 2.

As for your sample code: Why do you need two extensions in the first place? I don't really understand your example fully, but what you're doing in `NukkitExtension` seems to be test-case related, right? In that case I think it should rather be done in a before-each-method or the test method itself. Maybe `PowermockExtension` could provide the `PowermockState` object via the parameter resolution extension point (`MethodParameterResolver`)?

**Comment by jlink** [comment]
@johanhaleby Let's see if I understand you correctly...

You see two alternative (?) ways to implement PowerMock support:
1. Via deep copying the test instance using a factory as described in #203. You don't like this approach because it's bad performance-wise and error-prone.
2. Using a specialized class loader for ... yes for what parts of the test lifecycle exactly?
- Creating the test instance
- Running the test?
- Running all (or some) extension points? Creating the extension points?
- Would it be the same class loader for all tests or a tailor-made one for every test case?

Have I understood what you meant?

And yes, usually JUnit5 would drive PowerMock. That said, we could make the default class loader globally configurable. But I fear we'd be opening many cans of worms with that.

**Comment by mepeisen** [comment]
> Why do you need two extensions in the first place?
I personally do not need both. It is only a show case. I am fine with a single extension that does everything internally or with a meta extension that loads both in correct order. However this will be solved in final release it will be fine for me as Long as there is any solution for powermock :-)

> but what you're doing in  NukkitExtension  seems to be test-case related, right?
The Registration of instrumented classes (the lines 58 to 63) must be done before the new test instance is created. So that the class loader knows what to do. I decided to put it into postProcess because of your answer above that it may be changed for returning a new test instance. Maybe the powermock guys find another solution for it. There are already other annotations in powermock to specify instrumented classes on test method level. Simply do not know the powermock internals :-)

**Comment by johanhaleby** [comment]
@jlink:

1. Yes
1. Historically PowerMock has used either:
  1. One classloader to create the test instance and run all test cases in a class
  1. One classloader to create the test instance and run all test cases excepted those explicitly defined with `@PrepareForTest` that are executed in another classloader

Personally I would be fine with only supporting (i). What do you think @thekingnothing?

> And yes, usually JUnit5 would drive PowerMock. That said, we could make the default class loader globally configurable. But I fear we'd be opening many cans of worms with that.

I understand your concerns, but if not I guess PowerMock would have to do some hacks to try to launch JUnit with its classloader.

**Comment by mepeisen** [comment]
Instead opening class loader configuration there may be another smart solution: What about delegating not only the test instance creation but the extension creation to a meta-Extension?
What I mean is a two phase initialization for junit:
1) Look if there is a meta extension. If there is a meta extension ask it for test instance and for the Extension instances (maybe giving the classes to the meta extension).
2) If there is no meta extension create the test instance by itself and the extension instances by searching for @ExtendWith

This willl open a whole new bunch of use cases. For example a meta Extension that filters the extensions because of annotations. For example one might disable an Extension within Hudson (=env variable) but enable it within local eclipse.

And for Pockermock it enables nice Features, for example powermock can look for @PrepareForTest even on other Extension classes and not only on the test class itself.

**Comment by thekingnothing** [comment]
I think, ideally, it will be awesome if PowerMock could register anyhow global class loader (or class loader factory, which will be better, because PowerMock will still be able control class loader creating). The class loader will be used to create test instance, run test, load extension and so on. 

This approach will give a chance to us to discard checking if class should be deferred to another class loader (test/application class loader) to avoid class cast exception and code will be more simple and bug-free. 

By the way, @mepeisen it's an answer why we don't include "Mockito" classes to classes which always should be ignored by PowerMock class loader: some times it should be loaded via PowerMock class loader, because some internal classes is loaded only in case when you call `Mockito.mock`. In this case PowerMock class loader will be used to create these classes and if other Mockito classes is loaded via not PowerMock class loader then classes get incompatible. 

As I mentioned now, PowerMock has an option which lets a user to specify which classes should not be loaded by PowerMock classloader. It's useful for case, when classes have been loaded before PowerMock takes control test run process.

@mepeisen, 

PowerMock have to control all pointed stages: 

- Creating the test instance
- Running the test
- Running all extension points

to avoid class of problems which PowerMock has now with TestNG, when we can only control creating the test instance and partially controls running the test via setting context class loader. Some of them I mentioned above.

**Comment by thekingnothing** [comment]
@jlink, 

I also understand your doubts about giving such Powerful option which could breaks all in case improper using. But what do you think to allowed register class loader or class loader factory only from trusted source which directly allowed in code. Seems to me not a lot of test frameworks need such feature. Another way will be create a one new module **junit5-powermock**, which could be a bridge between jUnit and PowerMock and could be delivered with jUnit5. What do you think?

**Comment by thekingnothing** [comment]
@mepeisen, 

PowerMock have to control all pointed stages: 

- Creating the test instance
- Running the test
- Running all extension points

to avoid class of problems which PowerMock has now with TestNG, when we can only control creating the test instance and partially controls running the test via setting context class loader. Some of them I mentioned above.

**Comment by thekingnothing** [comment]
@johanhaleby, 

> Personally I would be fine with only supporting (i). What do you think @thekingnothing?

I think we can limit to use one class loader, but there are boundary cases when it could be useful. I mean load each test by a new classloader. 

For example: test uses a class which has a static (may even final field). The field is filled during static initialisation with using a mocked class which generate return value base on test conditionals. After first run, the class will be initialised and field will have correct mock value, when a second test runs field will still have value from previous test. The user cannot do anything to solve the problem without changing production code. 

Code: 

```
public class ClassA{

    private static final ClassB CLASS_B_INSTANCE = ClassBFactory.create();

    public String doSomething(){
       return  doSomethingOnBResult(CLASS_B_INSTANCE.message());
    }

    private String doSomethingOnBResult(String message){
      return "_" + message + "_";
    }
} 
```

```
public class ClassB {

  public String message(){
    return "Hello";
  }

}
```

```
public class ClassBFactory{

   private static ClassBFactoryImpl factoryImpl;
   
   public static void setImplementer(ClassBFactoryImpl factoryImpl){
     this.factoryImpl = factoryImpl;
   }

   public static ClassB create(){
     return factoryImpl.create();
   }
}
```

```
public class TestClassA{

   private ClassA instance;
   private ClassB clessbmock; 

  @Before
  public void setUp(){
     ClassBFactoryImpl factoryImpl = mock(ClassBFactoryImpl.class);
     ClassBFactory.setImplementer(factoryImpl);

     clessbmock = mock(ClassB.class);
     when(factoryImpl.create).thenReturn(clessbmock);

     instance = new ClassA();
  }

  @Test
  public void testCaseA(){
    when(clessbmock.message()).thenReturn("Welcome"); 
    
    String result = instance.doSomething(); // result will be correct "_ Welcome_"

  }

  @Test
  public void testCaseB(){
    when(clessbmock.message()).thenReturn("Hi"); 
    
    String result = instance.doSomething(); // result will be incorrect "_ Welcome_", because  instance still use old mock and mock cannot be replaced 
  }
}
```

I don't see a way to handle this case without reloading class with new class loader.

**Comment by jlink** [comment]
@thekingnothing @johanhaleby 
So do you think an extension point like that could suffice:
```
public interface TestInstanceClassLoaderProvider extends ExtensionPoint {
    ClassLoader provide(ContainerExtensionContext context);
}
```
The provided class loader would then be used to create a test instance and set as context class loader for the thread that is running test execution and all instance related extension points like `InstancePostProcessor`, `Before/AfterEachExtensionPoint` and `ExceptionHandlerExceptionPoint`.

And if that suited your needs, would the instance creation extension point from #203 still be needed?

**Comment by thekingnothing** [comment]
@jlink 

I think it's what we need and it will suffice most cases.  Will the  `TestInstanceClassLoaderProvider` be called once?

**Comment by thekingnothing** [comment]
And yes, in case if `TestInstanceClassLoaderProvider` is added, then #203 will be redundant.

**Comment by jlink** [comment]
@thekingnothing The way I imagined it, it would be called once per "container" i.e. test class or nested test class. But it could also be called freshly for each test instance.

**Comment by thekingnothing** [comment]
@jlink, I'd like to suggest the following semantic for the `provide` method. If it reruns `null`, then test container will continue use the current class loader. Otherwise the returned class loader will be used for all purposes. 

I still doubt if it's  really needed to be able to create a new class loader per test method. In rare cases it's useful, but but it more dangerous than it's worth.

**Comment by jlink** [comment]
Hm. Don't like `null` as some magic value. Moreover, the behaviour you suggest would require JUnit to store the last returned class loader. This is the kind of mutable state I'd rather avoid. 

If an extension wants to always return the same instance of class loader it can easily cache that instance internally (using a store from the context). Maybe `provide` needs additional information about what test case is going to use the class loader, e.g.:
`ClassLoader provide(ContainerExtensionContext context, Method testMethod);`

**Comment by thekingnothing** [comment]
I was thinking that jUnit anyway will store a class loader to use it load classes. It was my improper guessing.

I agree that an exception can care about class loader instance and it could be include into documentation for this exception. 

Hm, if exception will have information about method, it will be cool and PowerMock could still provide chunk feature. We have feature request to support running each test with a new class loader.

**Comment by thekingnothing** [comment]
Hi all, 

As I understand, this feature will not be implemented in junit 5.0. Am I right? It means that I'll not be able to provide supporting of junit 5 in PowerMock.

**Comment by deependratyagi** [comment]
When we can see support for powermock with junit 5?

**Comment by marcphilipp** [comment]
Not in 5.0, but possibly in 5.1 if we get a better understanding of your use case.

Would letting an extension create the test instance (as proposed in #672) also work?

**Comment by thekingnothing** [comment]
Maybe it could help. I'll check corner cases and recall why we came to conclusion that the given extension is better. 
```
public interface TestInstanceClassLoaderProvider extends ExtensionPoint {
    ClassLoader provide(ContainerExtensionContext context);
}
```

**Comment by AlexTrotsenko** [comment]
I see there is no activity on this issue for a while: any updates on `powermock` issue?

**Comment by marcphilipp** [comment]
@AlexTrotsenko I'm afraid not.

First, we need to have a clear understanding what extension point would be required.

@thekingnothing Have you had time to check on corner cases etc.?

**Comment by AlexTrotsenko** [comment]
@marcphilipp  Thanks for the update

**Comment by thekingnothing** [comment]
Hi,

There is no progress from PowerMock team regarding this issue.

The main goal for PowerMock - prepare 2.0 version which does not include support jUnit 5. However, due to lack of time of active contributors we will be able to release PowerMock 2.0 only in summer.  You may expect any progress in integration PowerMock and jUnit 5 not early than in Autumn.

On 19 Mar 2018, at 09:08, Alex Trotsenko <notifications@github.com<mailto:notifications@github.com>> wrote:


@marcphilipp<https://nam01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmarcphilipp&data=02%7C01%7C%7C13f192ab93194d43b2e908d58d70a414%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636570437282227552&sdata=Ohb%2FWcBQFEYJrTt8LqklQ2FvQhJtIo2v9HRr2jGYsvo%3D&reserved=0> Thanks for the update

**Comment by asarkar** [comment]
Unfortunately, the only ones at the receiving end of this "you first" tug of war are the users of both JUnit 5 and Powermock.

**Comment by marcphilipp** [comment]
@asarkar We really want to help, but (to cite myself):

> First, we need to have a clear understanding what extension point would be required.

**Comment by sbrannen** [comment]
FYI: I have renamed this issue to focus on the ability to provide a custom `ClassLoader`.

If there are other concerns specific to Powermock, I would suggest that new issues be opened to address those specific concerns.

**Comment by dpreussler** [comment]
just to bring the discussion from https://github.com/junit-team/junit5/issues/672 to here:

needed would be the way of loading the test class (and everything from there) by a different class loader.
This was very easily possible with Runners before, example:
https://github.com/dpreussler/kotlin-testrunner/blob/master/kotlin-runner-junit4/src/main/kotlin/de/jodamob/kotlin/testrunner/KotlinTestRunner.kt
(this project removes `finals` (the default in kotlin) for tests for example)

The new `TestInstanceFactory` does not offer the same functionality as the class itself is already loaded.

Next to Powermock the biggest user of this would probably be Robolectric: https://github.com/robolectric/robolectric

**Comment by markusheiden** [comment]
The TestInstanceFactory can load test class instances by a different class loader, but the check in ClassTestDescriptor#invokeTestInstanceFactory forces the created instance to be of the same class as the original test class. If this check would be softened e.g. to just check for the name of the test class, that would enable custom class loader support. Because the test instance is loaded by a different class loader, it should be no problem that the original test class already has been loaded by the original class loader.

**Comment by sbrannen** [comment]
@markusheiden, your claims are unfortunately in contradiction to my findings in https://github.com/junit-team/junit5/issues/672#issuecomment-410665795 (based on the status quo in `master`):

Or were you alluding to the possible "alternative" I brought up in https://github.com/junit-team/junit5/issues/672#issuecomment-410671753 ?

**Comment by markusheiden** [comment]
I did not notice that problem before. But yes, fixing the caching is an option. To fix the caching, one could add a guard checking the cached callback to be for the cached class and re-cache it, if it does not match. This would avoid performance decreases for the common case.

**Comment by JiangHongTiao** [comment]
Guys, any update regarding this? It would be very helpful to fully migrate to junit5 and not to keep old libraries in the product. Thanks for your time & effort!

**Comment by aviatix** [comment]
+1 for the idea to support junit5.

**Comment by bobtiernay-okta** [comment]
Would really love to remove the dependence on JUnit 4 in our codebase. This is the only lingering issue that is preventing that by way of our Powermock dependency. Would love to see a path forward here. Thanks for such a flexible library!

**Comment by Tibor17** [comment]
Do you have `SPI (Java Service Provider Interface)` to customize class loading in JUnit5?
As for instance such SPI can be a factory producing `ClassLoader` or test `Class` and similar, producing it depending on test context (CDI containers isolated by ClassLoader in your tests), consumed inside in the engine.

**Comment by marcphilipp** [comment]
Currently, there's `TestInstanceFactory` which should work for some of the use cases but it doesn't prevent "pre-loading" the test class with the current thread context classloader.

**Comment by Tibor17** [comment]
@marcphilipp 
JUnit5 code makes `pre-loading` the test classes somewhere?

**Comment by sormuras** [comment]
It depends on how the discover and/or execution request built by the JUnit Platform invoking tool is set up. If that tool only relies on basic default types, like `Path`, `File`, etc. test classes are only loaded just before test execution. If tools perform their own scanning and provide loaded test class instances, they are used "as-is".

**Comment by marcphilipp** [comment]
In the case of Jupiter, we use reflection to determine whether a class is a test class. Thus, we have to load the class before we can search for testable methods etc. That happens during test discovery, long before execution, and is unavoidable.

**Comment by marcphilipp** [comment]
@mannodermaus Could you please describe Robolectric's use case for this in more detail for us?

**Comment by Tibor17** [comment]
@sormuras 
and the last option which was not answered is that we provide `string` as fully qualified class name.
The Class loading is done on the fly with using the strings?
In this case with strings, how can we alter the ClassLoader? What is the mechanism to pass the ClassLoader implementation in to the `junit-platform-launcher`?

**Comment by sormuras** [comment]
At the moment, setting the [TCCL](https://github.com/junit-team/junit5/blob/master/junit-platform-commons/src/main/java/org/junit/platform/commons/util/ClassLoaderUtils.java#L39-L51) is the only option.

**Comment by sormuras** [comment]
An TCCL builder is implemented in https://github.com/sormuras/junit-platform-isolator -- which lacks documentation. That proof-of-concept project may serve a basis to solve this and other issues regarding custom `ClassLoader` instances.

**Comment by Tibor17** [comment]
@sormuras
I saw your project but overriding ClassLoader in the tool and by the tool is not the way we want to have. We want to have an extension API in JUnit5 which would use `our` ClassLoader from somewhere, i.e. IntelliJ IDEA or Maven plugin config. Overriding tool's ClassLoader via `thread.setContextClassLoader(loader)` is very risky.

**Comment by marcphilipp** [comment]
@Tibor17 Why is that more risky than setting the class loader using aa different API?

**Comment by Tibor17** [comment]
@marcphilipp 
If you was developer of IDEA or enterpreneur of Jetbrains, would you accept some external library to override the CL of IDEA and dedicate the control to that library? Nobody would accept this solution!

**Comment by marcphilipp** [comment]
@Tibor17 How would this allow setting the CL of IDEA? Besides, any test could set the TCCL anyway.

**Comment by sormuras** [comment]
Settting and forgetting TCCL -- see related issue #1799

**Comment by Tibor17** [comment]
#1799  is mockito related, feel free to discuss #1799 with @thekingnothing

**Comment by dpreussler** [comment]
@marcphilipp Robolectric is using custom class loader to intercept Android classes and replace them with "shadow" implementations that have basic functionality on JVM. This is needed as lot of classes can't be wrapped as they contain static methods or references that are only their on the Android VMs.
Similar use case I had in a library from the earlier Kotlin days: https://github.com/dpreussler/kotlin-testrunner where I needed to modify classes before testing.

As we discussed this for quite a while, see also i.e.
https://github.com/junit-team/junit5/issues/672
and the Android community is blocked on this and Google engineers and the Google testing team advocates against junit5 a lot, a solution would be appreciated.

@sormuras in my experience setting the thread class loader in the test or extension is too late and was confirmed by @sbrannen. The only way I got it working is to do this in a custom Engine. So any tips appreciated.

**Comment by Frontrider** [comment]
@dpreussler I'd be interested in that engine.

**Comment by sormuras** [comment]
With #157 solved by https://github.com/junit-team/junit5/commit/bb795075870ca1cd988d93fb359a500c72c0e67f ... you might be able to use the `InvocationInterceptor` extension API applied to your use-case. Not?

**Comment by marcphilipp** [comment]
@sormuras How so? The test class is already loaded when `InvocationInterceptors` are called, isn't it?

**Comment by wilkinsona** [comment]
This would be useful for Spring Boot's tests where we have a [custom runner](https://github.com/spring-projects/spring-boot/blob/5e62faf7201a8f929c052cab752c41da1b0deb22/spring-boot-project/spring-boot-tools/spring-boot-test-support/src/main/java/org/springframework/boot/testsupport/runner/classpath/ModifiedClassPathRunner.java) that allows tests to be run with a modified class path. I'd be happy to spend some time on contributing something if agreement can be reached on how the extension point should be implemented.

**Comment by Nilzor** [comment]
>Google engineers and the Google testing team advocates against junit5 a lot

Sorry for derailing but it would be interesting to read about why - do you have some source? If there are good reasons why I as an Android developer should stay away from Junit5, I'd like to know

**Comment by tasomaniac** [comment]
Afaik, that's only some Google engineers' personal opinions. Also, the Android testing team announced that they are working on their own solutions (project nitrogen). This was 2 years ago. Nothing public has been announced yet. 😔

Keep in mind that Android is unique in that sense. First are instrumentation tests running in device and those are mostly for UI tests. Google's tools for connected device tests are based on junit4. Secondly, unit tests are just built on top of Gradle's support and should very easily support Junit4 and 5.

In some chats, I have heard from Android Studio team that they are open to support jUnit 5 if there are issues preventing people from using it.

**Comment by dpreussler** [comment]
It's not only personal opinion. The Google testing team for Android decided not to support Junit5 as it's too much work for them. They hope though that community comes in.

There is no reason to avoid using it in your Android project as Gradle supports it really well.. Just keep in mind it won't work out of the box for UI tests. Robolectric is this weird hybrid in between and the biggest problem here.

**Comment by Frontrider** [comment]
Any updates on this one?

**Comment by marcphilipp** [comment]
We briefly discussed this issue during our last team call. We agreed someone (tm) we should pick a concrete use case and introduce the necessary extension points to make it work. However, work on this issue hasn't started, yet.

**Comment by ledoyen** [comment]
If it is of any help, here is  something I hope to do properly with the mentionned `ClassLoaderProvider` extension : https://github.com/fridujo/classpath-junit-extension

**Comment by vlsi** [comment]
One more use case: execute tests with custom bytecode transformation.

For instance: coverage-guided fuzzing with https://github.com/rohanpadhye/jqf

**Comment by filiphr** [comment]
Another concrete use case which is similar to the classpath-junit-extension is the [`ModifiedClassPathExtension`](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-tools/spring-boot-test-support/src/main/java/org/springframework/boot/testsupport/classpath/ModifiedClassPathExtension.java) which is used in the Spring Boot tests.

The way that this has been implemented allows to skip the initial execution via an `InvocationInterceptor` and then use the Launcher to launch the test again with the modified class loader. I have also went down this path with my requirement in our tests infrastructure. However, this adds ~300ms of execution time per test, which is understandable since it goes through the whole discovery scenario again.

I think that there are a lot of interested parties for this extension point and I think that everyone would be willing to try out in their own respective projects.

**Comment by jlink** [comment]
Does @junit-team have a bounty scheme? I’d be willing to tackle the issue with a reasonable incentive ;-)

**Comment by sormuras** [comment]
Do you have already a plan (solution) in mind, @jlink? Which use-case(s) would it solve?

@mannodermaus -- what's the status of https://github.com/mannodermaus/android-junit5? Any custom classloader required here, too?

**Comment by jlink** [comment]
@sormuras I was more or less joking but Powermock support shows up regularly on SO, thus I’d probably start with that if I were you.

**Comment by asarkar** [comment]
I'm curious how JUnit team prioritizes open issues. An issue that's been open for 4 years, gathered 42 upvotes, and has 23 watchers would seem to a layman a popular one to pick for an upcoming release. There's always the usual retort "if you want it, submit a PR yourself", but given the discussions, it didn't look like the core team even wanted to go down that route and welcome PRs.

**Comment by Frontrider** [comment]
Especially with an issue that is pretty much _**the**_ reason to keep using JUnit 4, over 5.

**Comment by mannodermaus** [comment]
@sormuras 
> what's the status of https://github.com/mannodermaus/android-junit5? Any custom classloader required here, too?

It's not required to function _per se_, however custom class loaders will alleviate one of the biggest current pain points of Android & JUnit 5, which is support for Robolectric tests. These still require a JUnit 4 environment and the Vintage Engine to be around in order to function properly. If an extension could bridge the gap between Robolectric's class loading shenanigans and the JUnit 5 world, the last reason not to switch to JUnit 5 for Android unit tests would be void at last.

**Comment by Frontrider** [comment]
People put so many hours of work into making JUnit5 as good as it can be (and it shows), and then we have this giant blocker sitting here for years, only people who forcefully want to use this create their (possibly expensive) workarounds.

At this point even a "works, but not well enough" solution could be a miracle.

**Comment by sbrannen** [comment]
Thank you everyone for the recent comments demonstrating your interest in this feature.

In light of that...

Tentatively slated for 5.7 M1 for _team discussion_

**Comment by sormuras** [comment]
fwiw, @thomasdarimont pointed me to [ClassLoader#getSystemClassLoader()](https://docs.oracle.com/javase/8/docs/api/java/lang/ClassLoader.html#getSystemClassLoader--) documentation, which contains a section that reads:

> If the system property `"java.system.class.loader"` is defined when this method is first invoked then the value of that property is taken to be the name of a class that will be returned as the system class loader. The class is loaded using the default system class loader and must define a public constructor that takes a single parameter of type ClassLoader which is used as the delegation parent. An instance is then created using this constructor with the default system class loader as the parameter. The resulting class loader is defined to be the system class loader.

Perhaps this `"java.system.class.loader"` system property can used already today to work around some problems described in this thread.

**Comment by thomasdarimont** [comment]
@sormuras thanks for the mention :) I initially stumbled upon this while inspecting the Source of [update4j's/DynamicClassLoader](https://github.com/update4j/update4j/blob/master/src/main/java/org/update4j/DynamicClassLoader.java)

**Comment by Vampire** [comment]
Mentioning here for the watchers that might prefer a solution, even if "slightly" awkward.
I posted a solution for using PowerMock with any JUnit platform based test, not only Jupiter tests, that leverages the JUnit 4 compatibility runner `JUnitRunner` provided by JUnit 5 at https://github.com/powermock/powermock/issues/830#issuecomment-616243661.

**Comment by Frontrider** [comment]
That sounds really nice and is a great workaround, but you should still try to put together something that would do this natively on the JunitPlatform.

I don't think you want to rely on an outdated version to do this for you.

**Comment by Vampire** [comment]
In no means I meant this as "that's the solution forever".
I just stumbled upon that runner and thought this could work and it does.
So I thought I share this with the people waiting for this integration as this issue just became 4 years old. :-D

**Comment by sbrannen** [comment]
4 years?  Wow..... has it really been that long? Time flies! :-o

**Comment by Frontrider** [comment]
It was there for 4 years, one has to make sure :)

**Comment by asarkar** [comment]
> 4 years? Wow..... has it really been that long? Time flies! :-o
Wow, we will be hard pressed to find another comment that implies more succinctly "Even though this issue has received 50+ votes and open for several years, we can't be bothered to care".

**Comment by Kricket** [comment]
Any chance of this being resolved?

Just as a side note, another case where this is needed is [Quarkus](https://quarkus.io/). The @QuarkusTest annotation loads the test class in a separate ClassLoader. We've had issues mixing this with Testcontainers annotations...

**Comment by zmeggyesi** [comment]
@jlink Has there been any progress on this issue? Is there any chance this will be implemented?

**Comment by asarkar** [comment]
74 upvotes on the original post, 28 watchers, and 4+ years later, still counting :)

**Comment by Yinjialiang** [comment]
oh, 4 years!  Time flies!!

**Comment by gayanW** [comment]
4 years, 7 months, 24 days, and counting.

**Comment by jlink** [comment]
Is anyone willing to put a worthwhile bounty on it? If so I’m willing to spent significant time on a solution. If not, then it’s more than ok for the JUnit team to spend their time on stuff which has more value for _them_.

**Comment by asarkar** [comment]
> Is anyone willing to put a worthwhile bounty on it?

Pretty sure buying labor isn’t how OSS is supposed to work. Good luck.

**Comment by jlink** [comment]
> > Is anyone willing to put a worthwhile bounty on it?
> 
> Pretty sure buying labor isn’t how OSS is supposed to work. Good luck.

You're absolutely right. It's mostly about people investing _their own private time_ at _their own discretion_.

**Comment by Frontrider** [comment]
There is actually a PR for this right now...

**Comment by dpreussler** [comment]
@Frontrider  Nice which one is that?

**Comment by filiphr** [comment]
I think that @Frontrider is referring to PR #2457. I've already mentioned in the PR, the PR goal is not to add an extension point for this issue. However, with it is possible to change the thread context class loader such that the classes can be loaded by it during the discovery phase. If however the discovery request is given already loaded classes then this won't work.

**Comment by markusheiden** [comment]
@jlink @asarkar I understand the frustration on both sides. 

The problem is, that JUnit 5 tries to replace JUnit 4 which is now already considered legacy. That is proved by the fact, that e.g. Spring Boot dropped support for it. But JUnit 5 is still not a complete replacement of JUnit 4, even after this many years. So those who need this feature are forced to use a "dead" library.

I hope this will encourage you (the JUnit 5 team) to consider this feature for one of the next releases. Anyway: Thanks for the work and providing a very nice piece of work!

**Comment by wilkinsona** [comment]
> That is proved by the fact, that e.g. Spring Boot dropped support for it

That's not the case. While the vintage engine is no longer part of `spring-boot-starter-test`, [you can add a dependency upon it and continue to test your Spring Boot applications using JUnit 4](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.4-Release-Notes#junit-5s-vintage-engine-removed-from-spring-boot-starter-test) if you wish to do so.

**Comment by jlink** [comment]
@markusheiden 

> The problem is, that JUnit 5 tries to replace JUnit 4 which is now already considered legacy. That is proved by the fact, 
> that e.g. Spring Boot dropped support for it. But JUnit 5 is still not a complete replacement of JUnit 4, even after this 
> many years. So those who need this feature are forced to use a "dead" library.

The implementation of Jupiter (especially of the testing lifecycle) differs enough from JUnit 4 that there will always remain some corner cases that are so rare and difficult that they'll never be tackled. The question at hand is if we have such a corner case here. 

I'd personally argue that Powermock's approach to do static mocking is sufficiently intrusive and demands so much collaboration from the testing framework that it might not be the most effective way to solve the testing problems for which it is being used. Given that Mockito meanwhile has a much less intrusive way to support some or even most of the use cases shows IMO that there are other options. I grant that Powermock is not the only tool that could make use of a custom ClassLoader but it's the one for which the simple approach ("just replace CL for test method execution") does not suffice. That probably means that this issue would long have been implemented and closed if it wasn't for this single tool's very special needs. I'd probably have "just done it" in 2016 when I was still part of the junit team.

**Comment by jlink** [comment]
> > That is proved by the fact, that e.g. Spring Boot dropped support for it
> 
> That's not the case. While the vintage engine is no longer part of `spring-boot-starter-test`, [you can add a dependency upon it and continue to test your Spring Boot applications using JUnit 4](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.4-Release-Notes#junit-5s-vintage-engine-removed-from-spring-boot-starter-test) if you wish to do so.

I really think it's a feasible option to have JUnit 4 engine (vintage) present just for the hopefully rare cases in which Powermock is the best tool. I regularly work in projects that use 2, 3 or even more test engines. It's not a hard burden.

**Comment by mannodermaus** [comment]
FWIW, this is not solely limited to only Powermock. One of the hardest challenges I face when trying to sell JUnit 5 to Android developers remains to be the fact that [Robolectric](https://github.com/robolectric/robolectric) cannot run on the JUnit Platform due to this ticket. The Vintage engine is great as a workaround, but ultimately a large chunk of the Android ecosystem is excluded from the advantages of upgrading, stuck in the duopoly of a JU4/5 hybrid that could be much improved if the Robolectric team had some means to intercept class loading.

**Comment by jlink** [comment]
@mannodermaus Fair enough. So, if Robolectric can get by with a simpler approach than Powermock, it might be possible to convince @junit-team/junit-committers to tackle that first. If I had an interest in Robolectric integration I'd probably research the necessary changes in Jupiter and create a new issue with just that requirement.

**Comment by marcphilipp** [comment]
+1 for a more focused issue with a concrete use case and suggestion how to achieve it.

**Comment by markusheiden** [comment]
> That is proved by the fact, that e.g. Spring Boot dropped support for it

Yes, that is technically wrong. What I wanted to express is, that JUnit 4 no longer automatically comes along as dependency with the Spring Boot test support.

> I really think it's a feasible option to have JUnit 4 engine (vintage) present just for the hopefully rare cases in which Powermock is the best tool. I regularly work in projects that use 2, 3 or even more test engines. It's not a hard burden.

Yes, you are right, it's not a big effort to achieve that. But the presence of JUnit 4 in the classpath provokes (static) imports (via IDE support) of JUnit 4 assertions in JUnit 5 tests, if one does not check twice on import. So for me it is a "big feature" not having JUnit 4 on the classpath in (mainly) JUnit 5 based projects.

**Comment by asarkar** [comment]
> That is proved by the fact, that e.g. Spring Boot dropped support for it

Even if it was true, which it isn’t, Spring Boot, while a popular web framework, is neither a goal nor a standard by which other frameworks are measured.

Of late, the discussion is veering again towards where it started half a decade ago, which is why the features asked for in this ticket are useful. At this time, it is clear that this ticket will not be implemented and every now and then someone will keep popping up with a new opinion notifying the many watchers unnecessarily. Why not just close the ticket as “won’t do” and be done with it instead of pretending that the ticket is being considered?

**Comment by marcphilipp** [comment]
> +1 for a more focused issue with a concrete use case and suggestion how to achieve it.

I'm not sure how much clearer I can express why this issue isn't resolved after "half a decade". I haven't given up hope that someone will come along with a concrete proposal instead of just complaining, yet. As for the many watchers getting notified unnecessarily, they can always unsubscribe should they no longer be interested.

**Comment by demiurg906** [comment]
One more use case: in Kotlin compiler we have a lot of independent tests which may be perfectly parallelized besides one fact: when compiler initializes it changes some static state and returns it back after test is finished. So if I just enable parallel test run than this state would be in inconsistent state. Perfect solution for this problem would be have separate classloaders for each thread which used for running tests

**Comment by TransLunarInjection** [comment]
Here is a concrete example of a test which worked on JUnit 4 and doesn't work with 5 and is due to classloading changes. I am not confident that it is a central example of the overall type of issue in this thread, but it seemed worth noting here due to earlier comments about a lack of a concrete use case.

https://github.com/MinimallyCorrect/JavaTransformer/pull/59

**Comment by Frontrider** [comment]
We have concrete use cases, but I'm not legally allowed to share them.

**Comment by ledoyen** [comment]
@TransLunarInjection the example you linked concerns the support of distinct classloaders for tests in junit-vintage-engine.

I think this is a different topic (linked but different), as the current issue concerns the existence of an API for extensions (such as a powermock one), ergo more in the junit-jupiter part.

On the orther hand, it could influence the design of a solution by moving its central parts from jupiter engine to the platform.

**Comment by gabizou** [comment]
A very unique use-case is being able to unit test that relies on class transformations via a custom class loader.

Some background: Sponge is an implementation of a Java API for Minecraft that uses its own tool ([Mixin](https://github.com/spongepowered/mixin)) to apply transformations on target classes (whether to implement the API interfaces, or add behaviors enabling event-based systems the API exposes), and while it'd be considered a massive integration test to "run various behaviors", we'd sooner wish to have better [JUnit-like tests](https://github.com/SpongePowered/SpongeForge/blob/ce6445b9c879b6765cc2ff91200b030d432986fe/src/integration-test/java/org/spongepowered/mod/test/integration/regression/BreakBlockUsedItem.java#L46) that could be run and verified for various API implementations that we'd better catch common bugs, and these would be better tested without having a full integration test being run.

A fair bit of the details about how the classes are transformed, we use [ModLauncher](https://github.com/McModLauncher/modlauncher) as our transformative classloader to apply such large sweeping changes.

**Comment by TWiStErRob** [comment]
:( sad to see this, I'm all for using JUnit 5, but until this is done, it's not viable to set up a fully Jupiter-based Android project.

**Comment by shaburov** [comment]
opened this issue on 13 Mar 2016 · 118 comments
**now** - 25 Aug 2022 

I wonder how many years to wait for a decision...

The author of the [related problem](https://github.com/robolectric/robolectric/issues/3397) already has grandchildren going to school...

**Comment by atsiporu** [comment]
Here is the link to the answer I got from @ledoyen 
https://github.com/junit-team/junit5/issues/3028#issuecomment-1309894679

**Comment by marcphilipp** [comment]
As it turns out, supporting PowerMock does not require any changes in JUnit if its existing Java agent is used along with a relatively simple Jupiter extension: https://github.com/powermock/powermock/pull/1146. To be frank, no new APIs were needed and this could have written years ago if someone had taken the time to give it a shot.

Besides the Java agent approach which allows transforming classes as necessary, the upcoming 5.10 release will introduce `LauncherInterceptor` (see #3091 for details) and there's an example for replacing the class loader for all tests in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/#launcher-api-launcher-interceptors-custom).

In #3028, we'll explore how to run tests with different classpaths. If you have a specific use case that you'd like to see supported that isn't covered by any of the above, please raise a new issue.

**Comment by jlink** [comment]
Good riddance
